Why min CWND of 2 instead of 1 (#3586)
quicwg/base-drafts: draft-ietf-quic-recovery.md

prior to a dramatic decrease in congestion window; see
{{persistent-congestion}}.

> ### Increase the min congestion window to 2 packets

Reviewer: Use "minimum congestion window" to stay consistent with rest of the draft. FWIW RFC 5681 for TCP refers to this as loss window LW. Not sure if we want to use the same term in this draft.

Reviewer 2: ```suggestion
### The Minimum Congestion Window is 2 Packets
```

Title Case is Used in This Document

> QUIC recommends a minimum congestion window of 2 packets instead of TCP's 1.

Reviewer 2: ```suggestion
QUIC recommends that the minimum congestion window be 2 packets instead 
of the 1 packet minimum in TCP.
```

2 packets avoid waiting for a delayed acknowledgement and allow the PTO to

> send 2 packets instead of 1, which can be particularly important during the

Reviewer 3: The PTO is not limited by the cwnd.

Reviewer 2: What Marten said.  This justification makes no sense.  Probing with two packets is different.  

As far as I can tell, this is an arbitrary choice, driven primarily by a desire to avoid the performance floor being too low.

Author: Always hitting delayed acks can be a major problem as well, though QUIC allows explicitly communicating those, which avoids some problems.

Reviewer 4: I see this as a more fundamental deviation to TCP than other changes, because although it will have some performance benefit in common cases, it also changes the flow aggressiveness in cases of extreme overload. As I udnerstand, the basis of TCP's back-off is that it scales back cwnd to 1 MSS, and then should scale back in the interval between sending 1 MSS (fractional MSS). If QUIC has a minimum of 2*MSS in QUIC, does it immediately change to 2*RTT interval, otherwise this appears to be significantly more aggressive flow under congestion collapse?

Author: I'd expect congestion collapse to be dominated by the PTO response, not the CWND.

I'm not aware of efforts to do fractional MSS in common use.  Do you have a reference for that?

Reviewer 4: Do we agree on what is intended by congestion-collapse, i.e. persistent long-lasting high level of congestion, as oppose to just "congestion"? (e.g. when there is a very low probability of a successful ACK). ... In will find something on reducing the MSS below once/RTT.

prior to a dramatic decrease in congestion window; see
{{persistent-congestion}}.

> ### The Minimum Congestion Window is 2 Packets

Reviewer 5: ```suggestion
### The Minimum Congestion Window is Two Packets
```

QUIC recommends that the minimum congestion window be 2 packets instead
of the 1 packet minimum in TCP. A minimum of 2 packets avoids waiting for a
delayed acknowledgement, which can substantially decrease throughput,

> particularly if the max_ack_delay is larger than the round trip time.

Reviewer 2: I'm not sure that this is true.  Hitting the minimum congestion window occurs after persistent congestion, at which point we might assume that any packet getting through will cause an immediate ACK (as there is likely a vast gap).  The only case where that doesn't happen is if only ACKs are being lost, where sending two packets don't help much as only the loss of the first will trigger an immediate ACK.

I think that the point below is stronger.

A 2 packet minimum congestion window also avoids waiting for a probe

> timeout(see {{pto}}) every time a single packet is lost.

Reviewer 2: ```suggestion
timeout (see {{pto}}) every time a single packet is lost.
```

> timeout(see {{pto}}) every time a single packet is lost.

Reviewer 2: I don't think that this is saying exactly what you want to say.

The point here, if I understand it, is that when an endpoint detects persistent congestion and reduces the window to the minimum, having two packets means that there is a far lower chance of BOTH of those packets being lost.  That greatly improves the likelihood that an acknowledgment will be returned and the connection can start to claw itself out of the hole it found itself in.

The point about sending a probe on PTO is only a consequence of the above; I wouldn't bother mentioning PTO.

> QUIC recommends that the minimum congestion window be 2 packets instead
> of the 1 packet minimum in TCP. A 2 packet minimum congestion window avoids
> waiting for a probe timeout (see {{pto}}) every time a single packet is lost.
> A minimum of 2 packets also avoids waiting for a delayed acknowledgement, which
> can substantially decrease throughput, particularly if the max_ack_delay is
> larger than the round trip time.

Reviewer 5: ```suggestion
TCP uses a minimum congestion window of one packet, but this can be fragile
in practice. Loss of a single packet sent with that window limits the sender to
waiting for a PTO ({{pto}}) to recover from that loss, which can be a
substantial period of time depending on the amount of backoff. Sending a
single data packet also increases the chances of incurring additional latency
when a receiver delays its acknowledgement. QUIC therefore recommends that
the minimum congestion window be two packets. While this increases network
load, it is considered safe, since the sender will still reduce its sending rate
exponentially under persistent congestion ({{pto}}).
```

prior to a dramatic decrease in congestion window; see
{{persistent-congestion}}.

### The Minimum Congestion Window is Two Packets

TCP uses a minimum congestion window of one packet. However, loss of

> a single packet sent with that window causes the sender to waiting for a PTO
> ({{pto}}) to recover from that loss, which is much longer than a round trip.

Reviewer 5: ```suggestion
that single packet means that the sender needs to waiting for a PTO
({{pto}}) to recover, which can be much longer than a round-trip time.
```
